### PR TITLE
Added entry points, and refactored and cleaned up the code

### PR DESCRIPTION
**Pull request by nhjjreuter**
My primary goal for this PR was to add entry points and refactor the code.

Entry points have been added (`htcb` and `htcrystalball`) pointing to `htcrystalball/main.py:main()`. The scripts were originally assumed to be called directly, which may break entry points hence I have had to move such functionality to the `main()` function. These changes also conveniently addressed issues #32, #27, and #31.

Since `config/slot_config.json` is a mutable file inside of the package directory, write permission errors will pop up if the package gets installed globally. If not, then users will modify each others' configurations on every run. I have provided a temporary fix by moving the slot config to `~/.htcrystalball` (i.e., the home folder of the current user). This is not the best solution and needs to be addressed, but in lieu of an alternative this will at least not break functionality.

Cleaning includes: 

- moving functions to different modules (i.e., `collect` for collecting data from htcondor, `examine` for examining the slot configuration with user inputs, and `utils` for non-specific utilities),
- removing redundant code (i.e., `check_dynamic_slots()`, `check_static_slots()`, and `check_gpu_slots()` contain a lot of duplicate code. Functionality can be preserved by merging into one function, now `check_slot_by_type()`), 
- Change variable names to be more consistent and removing redundant characters (i.e., `color_val` is no more informative than `color`, `arg_value` is not descriptive enough for a function argument name)
- Addressed some of the comments discussed with @jonamfischer that were not yet implemented (i.e., consistent usage of f-strings on some parts of the code)
- Modifying argparse arguments to contain default values, removing the need for the conditional statements in the entry point

Lastly, I have updated:

- `.gitignore` to cover a wider array of files that should not be pushed, addressing issue #24. 
- `_version.py` to make the package version available throughout the code (i.e., to allow `htcb --version`)
- `include_package_data` in `setup.py` by adding a `MANIFEST.in`
- versioning of the package requirements, fixing issue #23
- tests, because the refactoring broke the imports

### CODE REVIEW

**Comment by aqw** [comment]
Thanks for the PR. I'm still reading through the changes, but as a first request, please squash the `removed egg-info folder` and `removed .idea folder` up into `Refactoring and code cleanup, added entry points` to make it a bit more readable.

`git rebase -i HEAD~14` will start you off

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,2 @@
-rich
-htcondor
+rich>=6.1.1
+htcondor>=8.9.8
```
Comment (aqw) [inline requirements.txt:2]: This is an development release. Unless there's needed functionality in the 8.9.x series, I would prefer requiring `>=8.8.6`.

**Comment by aqw** [comment]
Just a quick recap of our convo:
- please cleanup some of the history with squashing
- reduce `.gitignore` rules
- look into test failures

Once that is addressed, then we'll get into the code itself more.

Since the PR is so large, my goal is to get it merged sooner rather than later, so others can open PRs without conflicting with this effort.

**Comment by jonamfischer** [comment]
@nhjjreuter  assert preview['sim_jobs'] <= ram_ratio results in 50 <= 0 which tells me there is something wrong in the calculation of the ram ratio

**Comment by nhjjreuter** [comment]
From what I have found so far, `ram_ratio = int(float(slot['ram']) / ram)` evaluates to `ram_ratio = int(5.0/10.0)`, `ram_ratio = int(0.5)`, `ram_ratio = 0`. Since I hadn't changed this test, I looked in the associated functions but could not find any changes that would break the original build. Could this be due to the contents of the slots_config.json file?

**Comment by nhjjreuter** [comment]
The bug seems to originate from the slots config file (with some slots having two `slot_size` entries). The order of these entries is not structured as it depends on how the htcondor collector returns them. It is 'fixed' if I manually remove the partitions from the slots config file, but there may be other factors at play here. It is worth looking into this.

One thing I can strongly recommend is to restrict your use of type conversions (i.e., float to int, int to str, etc.) unless necessary at the time of execution. Take into account that no rounding is done when converting a float to an int (i.e., `int(10.9) == 10`). The same applies for `f'{foo}:.2f'` (and by extension `'{0:.2f}'.format(foo)`). Use `int(round(foo, 0))` if you want to get rid of decimals.

**tl;dr** the build passes if the partitions are removed from the slot config. I'd recommend caution before assuming everything works as intended despite the tests succeeding.

**Comment by jonamfischer** [comment]
RE slot_size: since ram ratio and sim_jobs should be fetched from the same slot item in the list, they should be assigned correctly. What this error tells me is that preview['sim_jobs'] and ram_ratio don't correspond to one another. which makes sense as preview should contain multiple slots and the available ram should also be fetched from there so preview[slot_number]['sim_jobs'] <= req_ram/preview[slot_number]['ram_amount']

File: MANIFEST.in [hunk 2]
```diff
@@ -0,0 +1,3 @@
+include config/slots.json
```
Comment (aqw) [inline MANIFEST.in:1]: Is this file used by `setup.py`, or is an artifact from your local build process?

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,52 @@
...
+MANIFEST
+
+# Unit test / coverage reports
+.coverage
+.coverage.*
+.cache
+.pytest_cache/
+
+# Sphinx documentation
+docs/_build/
```
Comment (aqw) [inline .gitignore:43]: This can be removed. We don't use sphinx.

File: .gitignore [hunk 4]
```diff
@@ -1,3 +1,52 @@
...
+
+# Environments
```
Comment (aqw) [inline .gitignore:45]: These can be removed. They all depend on the user choosing to name the folder as such, and place it in the root of the repo.

File: setup.py [hunk 5]
```diff
@@ -33,16 +47,28 @@
...
-        'htcondor'
+        'rich>=6.1.1',
+        'htcondor>=8.8.6',
     ],
     tests_require=[
-        'pytest',
+        'pytest>=6.0.1',
     ],
-    data_files=[
-            ('share/man/man1', ['man/man1/htcrystalball.1', 'man/man1/fetch_condor_slots.1'])
```
Comment (aqw) [inline setup.py:46]: Do the man pages still install correctly with this removed?

File: setup.py [hunk 6]
```diff
@@ -1,22 +1,36 @@
...
+    re.M
+)
+
+if mo:
+    version = mo.group(1)
+else:
+    raise RuntimeError('Unable to find version string in %s.' % version_file)
+
+try:
+    import pypandoc
```
Comment (aqw) [inline setup.py:22]: Is there a motivation for doing this? According to the [setup.py docs](https://packaging.python.org/guides/making-a-pypi-friendly-readme/), Markdown is supported natively.

I'm curious what the advantage is to converting to .rst first. Otherwise, I don't see the advantage of introducing a large dependency.

**Review by aqw — APPROVED** [review]
Because of the moving mixed in with the changes, I was unable to review the changes to the main code itself.

I will review the diffs independently, and comment here if I find anything of significant concern.

As it stands, I have a few requests and changes, but nothing that stands in the way of merging. Let's work on this a bit and get it merged by this afternoon. :-)

File: setup.py [hunk 7]
```diff
@@ -1,22 +1,36 @@
...
+import re
 
-#
-# Grab some info from supporting files
-#
-with open('README.md', 'r') as fh:
-    long_description = fh.read()
+version_file = 'htcrystalball/_version.py'
+readme = opj(dirname(__file__), 'README.md')
+mo = re.search(
```
Comment (aqw) [inline setup.py:10]: What is this searching for?

File: htcrystalball/__init__.py [hunk 8]
```diff
@@ -1,7 +1,8 @@
 from ._version import __version__
 import logging
+from os.path import expanduser, join as opj
 
-SLOTS_CONFIGURATION = "~/.htcrystalball"
+SLOTS_CONFIGURATION = opj(expanduser('~'), '.htcrystalball')
```
Comment (aqw) [inline htcrystalball/__init__.py:7]: This entire commit (`Allow config file to be created if it does not exist`) can be squashed into the previous commit (`Change SLOTS_CONFIGURATION target`).

File: htcrystalball/collect.py [hunk 9]
```diff
@@ -26,7 +26,7 @@ def format_slots(slots: list) -> dict:
     for slot in slots:
         # TODO: Check if float conversion is necessary
         name = slot['UtsnameNodename']
-        n_cpus = int(slot.get('TotalSlotGPUs', 0))
+        n_cpus = int(slot.get('TotalSlotCpus', 0))
```
Comment (aqw) [inline htcrystalball/collect.py:30]: This commit can be squashed into `Set defaults for slot values`

File: htcrystalball/examine.py [hunk 10]
```diff
@@ -167,16 +167,20 @@ def default_preview(slot_name: str, slot_type: str) -> dict:
...
         'node': slot["UtsnameNodename"],
         'type': slot["SlotType"],
         'total_slots': slot["TotalSlots"],
         'cores': slot["TotalSlotCpus"],
-        'gpus': slot["TotalSlotGPUs"],
         'disk': slot["TotalSlotDisk"],
         'ram': slot["TotalSlotMemory"]
     }
 
+    if slot.get('TotalSlotGPUs', None):
```
Comment (aqw) [inline htcrystalball/examine.py:186]: Does htcondor not always return GPU slot information?

File: tests/test_crystal_ball.py [hunk 11]
```diff
@@ -1,36 +1,35 @@
 """Module for testing the htcrystalball module."""
-import os
+from htcrystalball import examine, collect, utils
+from os import path
 import argparse
 import pytest
-import htcrystalball as big_balls
-import fetch_slots as sloth
-import my_modules.check_slots as checker
+import json
```
Comment (aqw) [inline tests/test_crystal_ball.py:6]: As a comment (no action necessary), this commit (`Change test imports to reflect new module and function names `) should have probably been split into two, specifically because of the introduction of `json` and its usage later.

File: tests/test_crystal_ball.py [hunk 12]
```diff
@@ -1,8 +1,8 @@
 """Module for testing the htcrystalball module."""
 from htcrystalball import examine, collect, utils
 from os import path
+from pytest import fixture, raises as praises
```
Comment (aqw) [inline tests/test_crystal_ball.py:4]: I've seen this as a trend among python devs, so I assume it's best practice to limit the imports. But I've long enjoyed the namespacing of importing the entire module (easier to trace that way). Can you explain (or point to a discussion explaining) the benefits of reducing the scope of imports within a module?

**Review by aqw — APPROVED** [review]
This is another round of checks. Same as before. THere's a few questions and requests, but nothignt hat is a blocker.

There's a few additional commits that I would have liked to see squashed (`Include versions in requirements`, and `Reduce .gitignore`), but it's not worth it, because the git history between the origin and the fixup is complicated enough that it would be pain to get everything through. We'll leave them as is, but I did want to mention it.

File: requirements.txt (hunk 1 above)
Comment (aqw) [reply requirements.txt]: Fixed in a later commit now.

File: MANIFEST.in (hunk 2 above)
Comment (nhjjreuter) [reply MANIFEST.in:1]: Currently the package has only been cloned, so the setup is never run. If you run the setup, the package directly will only contain the files setuptools recognizes as necessary for the package to function, which _excludes_ data files. The `setup.py` up until now used `data_files`, but that is deprecated (and doesn't work with `wheels`). 

The `setup.py` also has the argument `include_package_data=True`, which checks the `MANIFEST.in` file for any files that should be included in the package directory when installed. However, there was no manifest file to check. Hence, I created a `MANIFEST.in` file that includes the data files, and removed the deprecated `data_files` argument.

File: .gitignore (hunk 3 above)
Comment (nhjjreuter) [reply .gitignore:43]: Will remove it, I figured that sphinx may be useful for future documentation (e.g., for readthedocs) so I didn't remove it yet. In any case, it can be re-added if we ever decide to use it.

File: htcrystalball/examine.py (hunk 10 above)
Comment (jonamfischer) [reply htcrystalball/examine.py:186]: Actually no, TotalSlotGPUs is only visible in slots that are children to a node with GPUs

File: MANIFEST.in (hunk 2 above)
Comment (aqw) [reply MANIFEST.in:1]: Thanks for the explanation. This is another example of a change that is self-contained and should be its own commit. And the description you provided above would be in the commit message.

Again, no need to change this one, but it should be kept in mind for future commits.

File: setup.py (hunk 6 above)
Comment (nhjjreuter) [reply setup.py:22]: Its actually not a dependency (it is used only if the package is already installed). The reason behind this is that some online resources (like PyPi) support rich text instead of markdown. If you think this is not necessary I will remove it.

File: setup.py (hunk 5 above)
Comment (aqw) [reply setup.py:46]: Resolving, as a result of the `MAINFEST.in` conversation above.

File: htcrystalball/examine.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+                int(total_ram / ram)
+            )
+        elif slot_type == 'gpu':
+            preview['sim_jobs'] = min(
+                int(total_gpus / n_gpu),
+                int(total_cpus / n_cpu),
+                int(total_ram / ram)
+            )
+        elif slot_type == 'static':
+            preview['sim_jobs'] = total_cpus
```
Comment (jonamfischer) [inline htcrystalball/examine.py:252]: @nhjjreuter Why do you have in here just the total_cpus for the number of similar jobs? This should be identical to the dynamic slot since the limits for similar jobs are the same (cpu cores and ram). 

Solution: if slot_type in ['dynamic', 'static']:
...

File: setup.py (hunk 7 above)
Comment (nhjjreuter) [reply setup.py:10]: It searches for the value of the `__version__` string in the `_version.py` file. Having a `_version.py` file makes it accessible from the code, rather than only in `setup.py`, which is advantageous if you need to inform the user of the package version (e.g., `htcb --version` returns the `__version___` value from `_version.py`).

File: htcrystalball/examine.py (hunk 10 above)
Comment (aqw) [reply htcrystalball/examine.py:186]: Ok. Thanks. Next time we have a few extra minutes to argue on Zoom, it would be nice to [bikeshed](https://en.wiktionary.org/wiki/bikeshedding) with everyone about our thoughts on the style here. Should there be a comment; `try` vs `if`.

File: setup.py (hunk 5 above)
Comment (nhjjreuter) [reply setup.py:46]: It should, through the manifest file. I'm not sure if this was ever tested through the setup, because the package has always been cloned, but not installed (correct me if I'm wrong). I'll check this out.

File: htcrystalball/examine.py (hunk 10 above)
Comment (nhjjreuter) [reply htcrystalball/examine.py:186]: I made this function work for all types of slots (e.g., dynamic, statis, gpu). If I recall correctly I got a key error when calling `TotalSlotGPUs` for a static or dynamic slot, but I'm not 100% sure. I will have to test it.

File: tests/test_crystal_ball.py (hunk 12 above)
Comment (nhjjreuter) [reply tests/test_crystal_ball.py:4]: There is no best practices here. Importing the entire package (vs. importing modules) is the same in terms of speed or what gets imported. The package is always imported in its entirety (although I'm sure there are exceptions, for example with packages that don't include all of their contents in the init file). For a big script I'd recommend going for `import pytest` and using `@pytest.fixture` and `pytest.raises` for readability and self documentation. For a small script such as this, especially if it is built around pytest, it shows precisely what functions from pytest we're using. In hindsight, I think `raises as praises` is not necessary (although I do appreciate the pun). It is simply a matter of style. For `from htcrystalball import examine, collect, utils` it confers a bigger benefit. We now know every module that is being tested.

A side-benefit is brevity in lines. While certainly not important (or something that necessitates these changes), `examine` is shorter than `htcrystalball.examine`, which can prevent awkward line-breaks at the 70 character limit. I would not recommend this for a big script, but for a small script like this the overview is not lost.

File: htcrystalball/examine.py (hunk 13 above)
Comment (nhjjreuter) [reply htcrystalball/examine.py:252]: Your original code has `preview_node['sim_jobs'] = available_slots`, where `available_slots = slot["TotalSlotCpus"]` in the `check_static_slots()` function (see line 169 in `my_modules/check_slots.py`.

I can make this change, but it may break the tests while functioning correctly.

Comment (jonamfischer) [reply htcrystalball/examine.py:252]: Ah, okay. Then the total_cpus confused me because available_slots = total_cpus is not always correct. Sometimes a single static slot has multiple cpus assigned, so I used the total slot number.

**Comment by aqw** [comment]
Going ahead and merging this. Additional fixes will follow in dedicated PRs. :-)
